dev/user.clj

```diff
@@ -2,7 +2,95 @@
   (:require [cheshire.core :as json]
             [clj-http.client :as http]
             [clj-momo.lib.time :as time]
-            [ctim.schemas.vocabularies :as vocab]))
+            [ctia.init :as init]
+            [ctia.properties :as p]
+            [ctim.schemas.vocabularies :as vocab]
+            [puppetlabs.trapperkeeper.app :as app]
+            [schema.core :as s]))
+
+(set! *warn-on-reflection* true)
+
+;;;;;;;;;;;;;;;;;;;;;;;
+;; Lifecycle management
+;;;;;;;;;;;;;;;;;;;;;;;
+
+;; To avoid losing the current system state, we manually
+;; intern a var in some namespace that is unlikely to be
+;; reloaded.
+
+;; initialize #'ctia.repl.no-reload/system-state
+(s/validate
+  {:app s/Any ;; nil or a running App
+   ;; coordination mechanism for #'serially-alter-app
+   :semaphore java.util.concurrent.Semaphore}
+  (-> (create-ns 'ctia.repl.no-reload)
+      (intern 'system-state)
+      (alter-var-root 
+        #(if (map? %)
+           ;; set root binding exactly once
+           %
+           {:app nil
+            :semaphore (java.util.concurrent.Semaphore. 1)}))))
+
+;; avoids #'locking and private vars to be robust in the presence of code reloading
+(defn serially-alter-app
+  "Alters the current app, except throws if more than 1 thread
+  attempts to alter it simultaneously."
+  [f & args]
+  (let [{^java.util.concurrent.Semaphore
+         s :semaphore} ctia.repl.no-reload/system-state
+        has-lock (.tryAcquire s)]
+    (try (if has-lock
+           (:app
+             (alter-var-root #'ctia.repl.no-reload/system-state
+                             ;; `constantly` to remove side effects
+                             (constantly
+                               (update ctia.repl.no-reload/system-state
+                                       :app #(apply f % args)))))
+           (throw (ex-info "Lifecycle management parallelism!"
+                           {})))
+         (finally
+           (when has-lock
+             (.release s))))))
+
+(defn current-app
+  "Returns the current app, or nil."
+  []
+  (:app ctia.repl.no-reload/system-state))
```

Comment:
`ctia.repl.no-reload/*` references might be slightly fragile? Because that ns cannot be required. Might confuse some tooling.

I'd favor `resolve` (maybe DRYed behind a defn)

Reply:
Great point, encapsulated.